File: vehicle-feature-catalog/src/python/vehicle_feature_catalog/indexes.py
"""In-memory indexes built once after load. Pure dict lookups thereafter."""

from __future__ import annotations

from dataclasses import dataclass, field

from .entities import Feature, Make, Model, Trim, TrimFeature
from .errors import DuplicateIdError


@dataclass
class CatalogIndexes:
    makes_by_id: dict[str, Make] = field(default_factory=dict)
    models_by_id: dict[str, Model] = field(default_factory=dict)
    trims_by_id: dict[str, Trim] = field(default_factory=dict)
    features_by_id: dict[str, Feature] = field(default_factory=dict)
    features_by_trim: dict[str, list[TrimFeature]] = field(default_factory=dict)
    trims_by_feature: dict[str, list[TrimFeature]] = field(default_factory=dict)
    models_by_make: dict[str, list[Model]] = field(default_factory=dict)
    trims_by_model: dict[str, list[Trim]] = field(default_factory=dict)

    @classmethod
    def build(
        cls,
        makes: list[Make],
        models: list[Model],
        trims: list[Trim],
        features: list[Feature],
        trim_features: list[TrimFeature],
    ) -> "CatalogIndexes":
        idx = cls()

        for m in makes:
            if m.id in idx.makes_by_id:
                raise DuplicateIdError(f"Duplicate Make id: {m.id}")
            idx.makes_by_id[m.id] = m

        for mo in models:
            if mo.id in idx.models_by_id:
                raise DuplicateIdError(f"Duplicate Model id: {mo.id}")
            idx.models_by_id[mo.id] = mo
            idx.models_by_make.setdefault(mo.make_id, []).append(mo)

        for t in trims:
            if t.id in idx.trims_by_id:
                raise DuplicateIdError(f"Duplicate Trim id: {t.id}")
            idx.trims_by_id[t.id] = t
            idx.trims_by_model.setdefault(t.model_id, []).append(t)

        for f in features:
            if f.id in idx.features_by_id:
                raise DuplicateIdError(f"Duplicate Feature id: {f.id}")
            idx.features_by_id[f.id] = f

        cls._check_cross_entity_id_collisions(idx)

        for tf in trim_features:
            idx.features_by_trim.setdefault(tf.trim_id, []).append(tf)
            idx.trims_by_feature.setdefault(tf.feature_id, []).append(tf)

        return idx

    @staticmethod
    def _check_cross_entity_id_collisions(idx: "CatalogIndexes") -> None:
        seen: dict[str, str] = {}
        for kind, bucket in (
            ("Make", idx.makes_by_id),
            ("Model", idx.models_by_id),
            ("Trim", idx.trims_by_id),
            ("Feature", idx.features_by_id),
        ):
            for entity_id in bucket:
                if entity_id in seen:
                    raise DuplicateIdError(
                        f"ID {entity_id!r} claimed by both {seen[entity_id]} and {kind}"
                    )
                seen[entity_id] = kind

File: vehicle-feature-catalog/src/ts/indexes.ts
import { Feature, Make, Model, Trim, TrimFeature } from './entities.js';
import { DuplicateIdError } from './errors.js';

export interface CatalogIndexes {
  makes_by_id: Map<string, Make>;
  models_by_id: Map<string, Model>;
  trims_by_id: Map<string, Trim>;
  features_by_id: Map<string, Feature>;
  features_by_trim: Map<string, TrimFeature[]>;
  trims_by_feature: Map<string, TrimFeature[]>;
  models_by_make: Map<string, Model[]>;
  trims_by_model: Map<string, Trim[]>;
}

export function buildIndexes(
  makes: Make[],
  models: Model[],
  trims: Trim[],
  features: Feature[],
  trimFeatures: TrimFeature[],
): CatalogIndexes {
  const idx: CatalogIndexes = {
    makes_by_id: new Map(),
    models_by_id: new Map(),
    trims_by_id: new Map(),
    features_by_id: new Map(),
    features_by_trim: new Map(),
    trims_by_feature: new Map(),
    models_by_make: new Map(),
    trims_by_model: new Map(),
  };

  for (const m of makes) {
    if (idx.makes_by_id.has(m.id)) {
      throw new DuplicateIdError(`Duplicate Make id: ${m.id}`);
    }
    idx.makes_by_id.set(m.id, m);
  }

  for (const mo of models) {
    if (idx.models_by_id.has(mo.id)) {
      throw new DuplicateIdError(`Duplicate Model id: ${mo.id}`);
    }
    idx.models_by_id.set(mo.id, mo);
    push(idx.models_by_make, mo.make_id, mo);
  }

  for (const t of trims) {
    if (idx.trims_by_id.has(t.id)) {
      throw new DuplicateIdError(`Duplicate Trim id: ${t.id}`);
    }
    idx.trims_by_id.set(t.id, t);
    push(idx.trims_by_model, t.model_id, t);
  }

  for (const f of features) {
    if (idx.features_by_id.has(f.id)) {
      throw new DuplicateIdError(`Duplicate Feature id: ${f.id}`);
    }
    idx.features_by_id.set(f.id, f);
  }

  checkCrossEntityCollisions(idx);

  for (const tf of trimFeatures) {
    push(idx.features_by_trim, tf.trim_id, tf);
    push(idx.trims_by_feature, tf.feature_id, tf);
  }

  return idx;
}

function push<K, V>(map: Map<K, V[]>, key: K, value: V): void {
  const list = map.get(key);
  if (list === undefined) {
    map.set(key, [value]);
  } else {
    list.push(value);
  }
}

function checkCrossEntityCollisions(idx: CatalogIndexes): void {
  const seen = new Map<string, string>();
  const buckets: Array<[string, Map<string, unknown>]> = [
    ['Make', idx.makes_by_id],
    ['Model', idx.models_by_id],
    ['Trim', idx.trims_by_id],
    ['Feature', idx.features_by_id],
  ];
  for (const [kind, bucket] of buckets) {
    for (const id of bucket.keys()) {
      const prior = seen.get(id);
      if (prior !== undefined) {
        throw new DuplicateIdError(
          `ID ${JSON.stringify(id)} claimed by both ${prior} and ${kind}`,
        );
      }
      seen.set(id, kind);
    }
  }
}
